File: Rust/2023/02.rs
#![feature(test)]

type Input = Vec<Game>;
type Game = Vec<CubeCounts>;

#[derive(Debug, Clone, Copy, Default)]
struct CubeCounts {
    red: u32,
    green: u32,
    blue: u32,
}

fn setup(input: &str) -> Input {
    input
        .lines()
        .map(|line| {
            line.split(": ")
                .nth(1)
                .unwrap()
                .split("; ")
                .map(|round| {
                    round
                        .split(", ")
                        .fold(CubeCounts::default(), |mut acc, round| {
                            let mut round = round.split(' ');
                            let count = round.next().unwrap().parse().unwrap();
                            let color = round.next().unwrap();
                            match color {
                                "red" => acc.red = count,
                                "green" => acc.green = count,
                                "blue" => acc.blue = count,
                                _ => panic!(),
                            }
                            acc
                        })
                })
                .collect()
        })
        .collect()
}

fn min_config(game: &Game) -> CubeCounts {
    game.iter()
        .fold(CubeCounts::default(), |acc, &round| CubeCounts {
            red: acc.red.max(round.red),
            green: acc.green.max(round.green),
            blue: acc.blue.max(round.blue),
        })
}

fn part1(input: &Input) -> usize {
    input
        .iter()
        .enumerate()
        .filter(|(_, game)| {
            let min = min_config(game);
            min.red <= 12 && min.green <= 13 && min.blue <= 14
        })
        .map(|(i, _)| i + 1)
        .sum()
}

fn part2(input: &Input) -> u32 {
    input
        .iter()
        .map(|game| {
            let min = min_config(game);
            min.red * min.green * min.blue
        })
        .sum()
}

aoc::main!(2023, 2, ex: 1);

File: Haskell/2023/02.hs
import Data.List
import Data.List.Split
import Lib

type CubeCounts = [Int]

type Game = [CubeCounts]

type Input = [Game]

main :: IO ()
main = aoc 2023 2 setup solve1 solve2 ["1"]

solve1 :: Input -> Int
solve1 = sum . map fst . filter (and . zipWith (>=) [12, 13, 14] . snd) . zip [1 ..] . map minConfig

solve2 :: Input -> Int
solve2 = sum . map (product . minConfig)

minConfig :: Game -> CubeCounts
minConfig = foldl1 (zipWith max)

setup :: String -> Input
setup = map parseGame . lines

parseGame :: String -> Game
parseGame = map parseCubeCounts . splitOn "; " . last . splitOn ": "

parseCubeCounts :: String -> CubeCounts
parseCubeCounts = collectCounts . map (fork (curry id) (read . head) last . words) . splitOn ", "

collectCounts :: [(Int, String)] -> CubeCounts
collectCounts = flip map ["red", "green", "blue"] . flip getCount

getCount :: String -> [(Int, String)] -> Int
getCount = (.) (maybe 0 fst) . (find . ((. snd) . (==)))
